libbpf-tools/statsnoop.c

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2021 Hengqi Chen
+//
+// Based on statsnoop(8) from BCC by Brendan Gregg.
+// 09-May-2021   Hengqi Chen   Created this.
+#include <argp.h>
+#include <errno.h>
+#include <signal.h>
+#include <time.h>
+
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "statsnoop.h"
+#include "statsnoop.skel.h"
+#include "trace_helpers.h"
+
+#define PERF_BUFFER_PAGES       16
+#define PERF_POLL_TIMEOUT_MS    100
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+static volatile sig_atomic_t canceled = 0;
+
+pid_t target_pid = 0;
+bool trace_failed_only = false;
+bool output_with_timestamp = false;
+static __u64 start_timestamp = 0;
+
+const char *argp_program_version = "statsnoop 0.1";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+const char argp_program_doc[] =
+"Trace stat syscalls.\n"
+"\n"
+"USAGE: statsnoop [-h] [-t] [-x] [-p PID]\n"
+"\n"
+"EXAMPLES:\n"
+"    statsnoop             # trace all stat syscalls\n"
+"    statsnoop -t          # include timestamps\n"
+"    statsnoop -x          # only show failed stats\n"
+"    statsnoop -p 1216     # only trace PID 1216\n";
+
+static const struct argp_option opts[] = {
+	{"pid", 'p', "PID", 0, "Process ID to trace"},
+	{"failed", 'x', NULL, 0, "Only show failed stats"},
+	{"timestamp", 't', NULL, 0, "Include timestamp on output"},
+	{NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help"},
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	long pid;
+
+	switch (key) {
+	case 'p':
+		errno = 0;
+		pid = strtol(arg, NULL, 10);
+		if (errno || pid <= 0) {
+			warn("Invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		target_pid = pid;
+		break;
+	case 'x':
+		trace_failed_only = true;
+		break;
+	case 't':
+		output_with_timestamp = true;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static void sig_int(int signo)
+{
+	canceled = 1;
+}
+
+void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	const struct event *e = data;
+	int fd, err;
+	double ts = 0.0;
+
+	if (e->ret >= 0) {
+		fd = e->ret;
+		err = 0;
+	} else {
+		fd = -1;
+		err = -e->ret;
+	}
+	if (!start_timestamp)
```

Comment:
start_timestamp is used only inside handle_event, let's make it static within that function only, instead of for the entire file